### PR TITLE
[Solidity] add support for the event keywords

### PR DESCRIPTION
**Pull request by prototypeC14**
This pull request adds support for the event keywords in Solidity.

Since the event only is used to record log, most of the time it will not affect the verification result. However, the parameter in the `emit` i.e. event call might have overflow issues.

### CODE REVIEW

File: src/solidity-frontend/solidity_convert.cpp [hunk 1]
```diff
@@ -4861,14 +4909,16 @@ void solidity_convertert::get_function_definition_name(
   //  - For function name, just use the ast_node["name"]
   // assume Solidity AST json object has "name" field, otherwise throws an exception in nlohmann::json
   std::string contract_name;
-  assert(ast_node.contains("scope"));
```
Comment (fbrausse) [inline src/solidity-frontend/solidity_convert.cpp:4864]: This `"scope"` hasn't been needed?

Comment (prototypeC14) [reply src/solidity-frontend/solidity_convert.cpp:4864]: Yes, it was used for debugging, but it is no longer needed.

File: src/solidity-frontend/solidity_convert.cpp [hunk 2]
```diff
@@ -963,6 +979,14 @@ bool solidity_convertert::get_function_definition(
 
     added_symbol.value = body_exprt;
   }
+  else
+  {
+    //! assume it's event-definition
+    // construct an empty body
+    assert(ast_node["nodeType"] == "EventDefinition");
```
Comment (lucasccordeiro) [inline src/solidity-frontend/solidity_convert.cpp]: Shall we use our log system here?

File: src/solidity-frontend/solidity_convert.cpp [hunk 3]
```diff
@@ -1483,6 +1507,15 @@ bool solidity_convertert::get_statement(
 
     break;
   }
+  case SolidityGrammar::StatementT::EmitStatement:
+  {
+    // treat emit as function call
+    assert(stmt.contains("eventCall"));
```
Comment (lucasccordeiro) [inline src/solidity-frontend/solidity_convert.cpp]: Shall we use our log system here?

**Review by lucasccordeiro — CHANGES_REQUESTED** [review]

File: src/solidity-frontend/solidity_convert.cpp (hunk 2 above)
Comment (prototypeC14) [reply src/solidity-frontend/solidity_convert.cpp]: Yes, I will add a log_error here.

File: src/solidity-frontend/solidity_convert.cpp (hunk 3 above)
Comment (prototypeC14) [reply src/solidity-frontend/solidity_convert.cpp]: Yes, I will add a log_error here.

File: src/solidity-frontend/solidity_convert.cpp (hunk 2 above)
Comment (lucasccordeiro) [reply src/solidity-frontend/solidity_convert.cpp]: Thank you.

File: src/solidity-frontend/solidity_grammar.cpp [hunk 4]
```diff
@@ -1017,7 +1026,9 @@ const char *var_decl_statement_to_str(VarDeclStmtT type)
 // auxiliary type to convert function call
 FunctionDeclRefT get_func_decl_ref_t(const nlohmann::json &decl)
 {
-  assert(decl["nodeType"] == "FunctionDefinition");
+  assert(
```
Comment (lucasccordeiro) [inline src/solidity-frontend/solidity_grammar.cpp:1029]: Shall we use our log system here?

File: src/solidity-frontend/solidity_convert.cpp [hunk 5]
```diff
@@ -3369,15 +3407,23 @@ bool solidity_convertert::get_func_decl_ref(
 {
   // Function to configure new_expr that has a +ve referenced id, referring to a function declaration
   // This allow to get func symbol before we add it to the symbol table
-  assert(decl["nodeType"] == "FunctionDefinition");
+  assert(
```
Comment (lucasccordeiro) [inline src/solidity-frontend/solidity_convert.cpp:3423]: Shall we use our log system here?

**Comment by ChenfengWei0** [comment]
Shall we merge this PR?

**Review by lucasccordeiro — APPROVED** [review]

**Comment by lucasccordeiro** [comment]
Thanks for submitting this PR, @prototypeC14 and @ChenfengWei0.

**Pull request merged**